DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/EncodedBlocks.h

```diff
@@ -808,59 +815,79 @@ void EncodedBlocks<H, N, W>::encode(const S_IT srcBegin,     // iterator begin o
-    dataSize = SizeEstMarginAbs + int(SizeEstMarginRel * (dataSize / sizeof(W))) + (sizeof(STYP) < sizeof(W)); // size in words of output stream
-    expandStorage(dictSize + dataSize);
+    dataSize = SizeEstMarginAbs + int(SizeEstMarginRel * (dataSize / sizeof(W))) + (sizeof(input_t) < sizeof(W)); // size in words of output stream
+    expandStorage(frequencyTable.size() + dataSize);
     //store dictionary first
-    if (dictSize) {
-      bl->storeDict(dictSize, frequencies->data());
+    if (frequencyTable.size()) {
+      thisBlock->storeDict(frequencyTable.size(), frequencyTable.data());
     }
     // vector of incompressible literal symbols
-    std::vector<STYP> literals;
+    std::vector<input_t> literals;
     // directly encode source message into block buffer.
-    auto blIn = bl->getCreateData();
-    auto frSize = bl->registry->getFreeSize(); // note: "this" might be not valid after expandStorage call!!!
-    const auto encodedMessageEnd = encoder->process(srcBegin, srcEnd, blIn, literals);
-    rans::utils::checkBounds(encodedMessageEnd, blIn + frSize);
-    dataSize = encodedMessageEnd - bl->getData();
-    bl->setNData(dataSize);
-    bl->realignBlock();
+    storageBuffer_t* const blockBufferBegin = thisBlock->getCreateData();
+    const size_t maxBufferSize = thisBlock->registry->getFreeSize(); // note: "this" might be not valid after expandStorage call!!!
+    const auto encodedMessageEnd = encoder->process(srcBegin, srcEnd, blockBufferBegin, literals);
+    rans::utils::checkBounds(encodedMessageEnd, blockBufferBegin + maxBufferSize);
+    dataSize = encodedMessageEnd - thisBlock->getData();
+    thisBlock->setNData(dataSize);
+    thisBlock->realignBlock();
     // update the size claimed by encode message directly inside the block
-    // store incompressible symbols if any
 
-    int literalSize = 0;
-    if (literals.size()) {
-      literalSize = (literals.size() * sizeof(STYP)) / sizeof(stream_t) + (sizeof(STYP) < sizeof(stream_t));
-      expandStorage(literalSize);
-      bl->storeLiterals(literalSize, reinterpret_cast<const stream_t*>(literals.data()));
-    }
-    *meta = Metadata{messageLength, literals.size(), sizeof(uint64_t), sizeof(stream_t), static_cast<uint8_t>(encoder->getSymbolTablePrecision()), opt,
-                     encoder->getMinSymbol(), encoder->getMaxSymbol(), dictSize, dataSize, literalSize};
+    // store incompressible symbols if any
+    const size_t nLiteralSymbols = [&]() {
+      const size_t nSymbols = literals.size();
+      if (!literals.empty()) {
+        // introduce padding in case literals don't align;
+        const size_t nPaddingElements = (nSymbols * sizeof(input_t)) % sizeof(storageBuffer_t);
+        literals.insert(literals.end(), nPaddingElements, {});
+
+        const size_t literalStorageSize = (literals.size() * sizeof(input_t)) / sizeof(storageBuffer_t) + (sizeof(input_t) < sizeof(storageBuffer_t));
```

Comment:
Same comment as before: here for `sizeof(input_t) < sizeof(storageBuffer_t)` you are always adding 1, while it should be added only if `(literals.size() * sizeof(input_t)) % sizeof(storageBuffer_t) != 0`.
See how I did this in https://github.com/AliceO2Group/AliceO2/blob/1e054d67466173870d49c8726f955f90de4f4f97/DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/EncodedBlocks.h#L848-L854